core/gate.cc

```diff
@@ -42,67 +43,92 @@ namespace bess {
 
   if (reset) {
-    all_gate_hook_factories.clear();
+    all_gate_hook_builders.clear();
   }
 
-  return all_gate_hook_factories;
+  return all_gate_hook_builders;
 }
 
-const std::map<std::string, GateHookFactory>
-    &GateHookFactory::all_gate_hook_factories() {
-  return all_gate_hook_factories_holder();
+const std::map<std::string, GateHookBuilder>
+    &GateHookBuilder::all_gate_hook_builders() {
+  return all_gate_hook_builders_holder();
 }
 
-// Creates new gate hook from the given factory, initializes it,
+const std::string Gate::GenerateDefaultName(const GateHookBuilder *builder,
+    Gate *gate) {
+  const std::string &name_template = builder->name_template();
+
+  for (int i = 0;; i++) {
+    const std::vector<GateHook *> hooks = gate->hooks();
+    std::string name = name_template + std::to_string(i);
+    auto it = std::find_if(hooks.begin(), hooks.end(),
+        CompareGatehookName(name));
+    if (it != hooks.end())
+      continue;
+    return name;
+  }
+}
+
+// Creates new gate hook from the given builder, initializes it,
 // and adds it to the given gate.  If any of these steps go wrong
 // the gatehook instance is destroyed and we return a failed
 // CommandResponse, otherwise we return a successful one.
-CommandResponse Gate::NewGateHook(const GateHookFactory *factory, Gate *gate,
-                                  bool is_igate,
-                                  const google::protobuf::Any &arg) {
-  bess::GateHook *hook = factory->hook_constructor_();
-  CommandResponse init_ret = factory->hook_init_func_(hook, gate, arg);
+GateHook *Gate::CreateGateHook(const GateHookBuilder *builder, Gate *gate,
+                               const std::string &name,
+                               const google::protobuf::Any &arg,
+                               pb_error_t *error) {
+  bess::GateHook *hook = builder->hook_constructor_();
+  CommandResponse init_ret = builder->hook_init_func_(hook, gate, arg);
   if (init_ret.error().code() != 0) {
     delete hook;
-    return init_ret;
+    *error = init_ret.error();
+    return nullptr;
+  }
+
+  std::string hook_name = name;
+  if (hook_name == "") {
+    hook_name = GenerateDefaultName(builder, gate);
   }
-  hook->set_factory(factory);
+
+  hook->set_class_name(builder->class_name());
+  hook->set_name(hook_name);
+  hook->set_builder(builder);
   hook->set_arg(arg);
   hook->set_gate(gate);
-  int ret = gate->AddHook(hook);
+  int ret = gate->AddHook(hook, error);
   if (ret != 0) {
     delete hook;
-    return CommandFailure(ret, "Unable to add hook '%s' to '%s' %cgate '%hu'",
-                          factory->hook_name_.c_str(),
-                          gate->module()->name().c_str(), is_igate ? 'i' : 'o',
-                          gate->gate_idx());
+    return nullptr;
   }
-  return CommandSuccess();
+  return hook;
 }
 
-int Gate::AddHook(GateHook *hook) {
+int Gate::AddHook(GateHook *hook, pb_error_t *error) {
```

Comment:
I think it is fine to keep setting `error` by the caller, to minimize the footprint of Protobuf-related code.